File: 1396-design-underground-system/1396-design-underground-system.java
class UndergroundSystem {
    
    Map<Integer, ArrivalInfo> arrivals;
    Map<String, double[]> info;

    public UndergroundSystem() {
        arrivals = new HashMap<>();
        info = new HashMap<>();
    }
    
    public void checkIn(int id, String stationName, int t) {
        arrivals.put(id, new ArrivalInfo(id, stationName, t));
    }
    
    public void checkOut(int id, String stationName, int t) {
        ArrivalInfo arrival = arrivals.get(id);
        
        String key = arrival.stationName + "_" + stationName;
        
        double[] pair = info.getOrDefault(key, new double[2]);
        
        int time = t - arrival.t;
        pair[0] += time;
        pair[1]++;
        
        info.put(key, pair);
    }
    
    public double getAverageTime(String startStation, String endStation) {
        
        String key = startStation + "_" + endStation;
        double[] data = info.get(key);
        
        return data[0] / data[1];
    }
    
    class ArrivalInfo{
        int id;
        String stationName;
        int t;
        
        ArrivalInfo(int id, String stationName, int t){
            this.id = id;
            this.stationName = stationName;
            this.t = t;
        }
    }
}

/**
 * Your UndergroundSystem object will be instantiated and called as such:
 * UndergroundSystem obj = new UndergroundSystem();
 * obj.checkIn(id,stationName,t);
 * obj.checkOut(id,stationName,t);
 * double param_3 = obj.getAverageTime(startStation,endStation);
 */

File: 1396-design-underground-system/1396-design-underground-system.cpp
class UndergroundSystem {
public:
    unordered_map<int, pair<string, int>> arrivals;
    unordered_map<string, pair<double, double>> info;
    
    UndergroundSystem() {
        
    }
    
    void checkIn(int id, string stationName, int t) {
        
        arrivals[id].first = stationName;
        arrivals[id].second = t;
    }
    
    void checkOut(int id, string stationName, int t) {
        
        auto arrival = arrivals[id];
        
        string key = arrival.first + "-" + stationName;
        
        int travelTime = t - arrival.second;
        
        info[key].first += travelTime;
        info[key].second++;
    }
    
    double getAverageTime(string startStation, string endStation) {
        
        string key = startStation + "-" + endStation;
        return info[key].first / info[key].second;
    }
};

/**
 * Your UndergroundSystem object will be instantiated and called as such:
 * UndergroundSystem* obj = new UndergroundSystem();
 * obj->checkIn(id,stationName,t);
 * obj->checkOut(id,stationName,t);
 * double param_3 = obj->getAverageTime(startStation,endStation);
 */